README.md
# pick_json
> Cross-platform command line utility to extract and/or evaluate json. 
> No weird new language or syntax; just use basic javascript notation in your one-liners. 
> Pipe data into it as part of script chain or read from a file. Honey Badger doesn't care.

## Options
```
  Usage: pick_json [options]

  Options:

    -h, --help              output usage information
    -V, --version           output the version number
    -k, --keys              Just output the keys. Will output the root keys if no expression is given
    -e, --exp <expression>  Expression to filter the json. Must start with an attribute or index
    -f, --file <file>       Use <file> instead of standard input
    -v, --verbose           Verbose errors

Example
    $ echo '[ { "bar" : 42 } ]' |  pick_json -e "[0].bar > 40" #returns true
```

## Installation
Assumes you have Node installed
```
npm install -g pick_json
```

## Examples

We are using the following json data file as input data in the examples. You can find it in the `test` directory
```javascript
{
   "redis": {
      "connected": true,
      "ready": true,
      "connections": 1,
      "commandsSent": 9,
      "commandQueue": {
         "tail": [],
         "head": [],
         "offset": 0
      }
   },
   "error_codes" : [2001, 4004]
}
```

### Picking from an array
```
pick_json -e error_codes[1] test/example.json
```
 returns `4004`

### Picking from a nested structure
```
pick_json -e redis.connected test/example.json
```
returns `true`

### Filtering data
Remember, the `expr` can be *any* valid expression operating on the data
```
pick_json -e "error_data.filter(err => err > 3000)" data.json
```
returns `[4004]`

### When receiving an array
```
echo [ { "bar" : 42 } ] |  pick_cli [0].bar 
```
returns `42`

## Extracting data from a http service
Use curl to extract the data and pipe it into `pick_json`

```
curl -s mysite.com/service.json | pick_json -e stats.uptime 
```



History.md
v2.0.2 / 2016-02-17
============================

  * Fix Windows CLI problem preventing global install

v2.0.1 / 2016-02-17
===================

  * Fix runtime error when specifying --verbose

v2.0.0 / 2016-02-17
===================

  * New incompatible interface
  * Use OS independant mechanism for reading stdin
  * Output error when used as in version 1
  * Add test for keys on sub elements

v1.5.0 / 2016-02-15
===================

  * Simplify: only handle array indices specially
  * Make preceding dots optional for arrays
  * Add array transform test for documentation

v1.4.0 / 2016-02-09
===================

  * Update tests and README
  * Auto-detect arrays
  * Add CLI tests using Bats
  * Fix a bug in the error handling


1.3.2 / 2016-01-19
==================

  * Pretty-print results

1.3.1
==================
  * Change examples for normal file access

1.3.0
==================
  * Add support for explicit file argument

1.2.0
==================
  * Add option to just list keys

1.1.2
==================
  * Error fix for nested variables

1.1.1
==================
  * Fix README and dependencies

1.1.0
==================
  * Added array support

1.0.1 / 2015-09-16
==================
  * README with examples
  * First version of cli util `pick_json`


test/test-driver.sh
DIR=$(dirname "$0")
CMD="node $DIR/../pick_json.js"

TEST1=$($CMD redis.connected < $DIR/example.json)
TEST2=$(echo '[ { "bar" : 42 } ]' |  $CMD -a [0].bar)

[[ "$TEST1" == "true" ]] \
    && [[ "$TEST2" == "42" ]]


pick_json.js
#!/usr/bin/env node --harmony
/* Evil eval based utility. Only pass me nice data ... */
'use strict';

const fs = require("fs");
const program = require('commander');
const packageJson = JSON.parse(fs.readFileSync(__dirname + '/package.json').toString());
const colors = require('colors');
const util = require('util');
let readable = process.stdin;
let stringData = "";
let objectExpression;
let evalString;
let json;
let result;
let isArray;

program
.version(packageJson.version)
.option('-k, --keys', 'Just output the keys. Will output the root keys if no expression is given')
.option('-e, --exp <expression>', 'Expression to filter the json. Must start with an attribute or index')
.option('-f, --file <file>', 'Use <file> instead of standard input')
.option('-v, --verbose', 'Verbose errors')
.on('--help', () => {
    console.log('Example\n    $ echo \'[ { "bar" : 42 } ]\' |  pick_json -e "[0].bar > 40" #returns true');
})
.parse(process.argv);

if (!program.keys && (typeof program.exp=== 'undefined')) {
    console.error('no expression given!');
    program.outputHelp(make_red);
    process.exit(1);
}

if(program.args.length) {
    outputError(`You supplied arguments that were not prefixed by an option flag: "${program.args}". Aborting.`);
    process.exit(1);
}

if(program.file) {
    readable = fs.createReadStream(program.file);
}

readable.setEncoding('utf8');

readable.on('data', (chunk) => {
      debugOutput(`got ${chunk.length} bytes of data\n`);
      stringData += chunk;
});

readable.on('end', () => {
    evaluateJSON(stringData);
});

readable.on('error', (err) => {
    outputError('Error occurred on reading data', err);
    process.exit(1);
});


function make_red (txt) {
    return colors.red(txt); //display the help text in red on the console
}

function outputError(txt, error){
    console.error(make_red(txt));
    if(error && program.verbose) {
        console.log(error.stack);
    }
}

// Do not output red colored output. Simply avoid mixing the output with the stdout
function debugOutput(txt) {
    if(program.verbose){
        process.stderr.write('> ' + txt);
    }
}

function evaluateJSON(stringData) {
    try {
        json = JSON.parse(stringData);
        isArray = Array.isArray(json);
    } catch (ex) {
        outputError(util.format(
            'Could not parse supplied JSON from %s: %s', 
            readable === process.stdin ? 'stdin' : readable.path,
            ex.message), ex);
            process.exit(1)
    }

    objectExpression = program.exp;
    try{

        if (!objectExpression) {
            evalString = 'json';
        } else if (objectExpression.match(/^\[\d+\]/)) {
            evalString = 'json'+objectExpression;
        } else {

            // This complex matching allows for evaluation of arbitrary expressions
            // i.e. "servers.filter( name => name == 'redis' )"
            let rangeAlpha = 'a-zA-Z';
            let rangeAlphaNum = `${rangeAlpha}0-9`
            let regex = new RegExp(`\\.?([${rangeAlpha}_][${rangeAlphaNum}_]*)(.*)`);
            let match = objectExpression.match(regex);

            let firstPart = match[1];
            let rest = match[2];
            let str = ( 'json["' + firstPart + '"]' + rest);
            evalString = str;
        }
        result = eval( evalString );

    } catch(err){
        outputError(`Failed processing "${objectExpression}"`, err);
        if (isArray) { console.log('Is the expression applicable to an array?'); }
        debugOutput(`String we tried to evaluate: ${evalString}`);
        process.exit(1);
    }

    if (result) {
        if(program.keys){
            console.log(Object.keys(result).join('\n'));
        } else { 
            console.log(JSON.stringify(result, null, 4)); 
        }
    } else {
        outputError('No data found using identifier ' + objectExpression);
        process.exit(1);
    }
}

